api/bibliotheca.py

```diff
@@ -303,6 +314,94 @@ def apply_circulation_information_to_licensepool(self, circ, pool, analytics=Non
+            cls, license_pool, findaway_license
+    ):
+        """Convert a Findaway license document to a standard Web Publication
+        Manifest (audiobook flavor).
+
+        :param license_pool: A LicensePool for the title in question.
+        This will be used to fill in basic bibliographic information.
+
+        :param findaway_license: A string containing a Findaway
+           license document, or a dictionary representing such a
+           document loaded into JSON form.
+        """
+        if isinstance(findaway_license, basestring):
+            findaway_license = json.loads(findaway_license)
+
+        context_with_extension = [
+            "http://readium.org/webpub/default.jsonld",
+            {"findaway" : cls.FINDAWAY_EXTENSION_CONTEXT},
+        ]
+
+        manifest = AudiobookManifest(context=context_with_extension)
+
+        # Add basic bibliographic information (identifier, title,
+        # cover link) to the manifest based on our existing knowledge
+        # of the LicensePool and its Work.
+        manifest.update_bibliographic_metadata(license_pool)
+
+        # Add Findaway-specific information as extra metadata.
+        for findaway_extension in [
+                'accountId', 'checkoutId', 'fulfillmentId', 'licenseId',
+                'sessionKey'
+        ]:
+            value = findaway_license.get(findaway_extension, None)
+            output_key = 'findaway:' + findaway_extension
+            manifest.metadata[output_key] = value
+
+        # Add the spine items. All of them are in the same format.
+        # None of them will have working 'href' fields -- it's just to
+        # give the client a picture of the structure of the timeline.
+        audio_format = findaway_license.get('format')
+        if audio_format == 'MP3':
+            part_media_type = Representation.MP3_MEDIA_TYPE
+        else:
+            logging.error("Unknown Findaway audio format encountered: %s",
+                          audio_format)
+
+        # TODO: The items are in an ordered list, but each one also
+        # has an explicit 'sequence'. For now we'll pass it on but
+        # assume that it is redundant and the ordered list is always
+        # correct.
+        #
+        # TODO: Each item has a 'part' which always seems to be zero.
+        # Its purpose is unknown. For now, we simply pass it on.
+        total_duration = 0
+        for part in findaway_license.get('items'):
+            title = part.get('title')
+
+            # TODO: Incoming duration appears to be measured in
+            # milliseconds. This assumption makes our example
+            # audiobook take about 7.9 hours, and no other reasonable
+            # assumption is in the right order of magnitude. But this
+            # needs to be explicitly verified.
+            duration = part.get('duration', 0) / 1000.0
+
+            kwargs = {}
+
+            sequence = part.get('sequence')
+            if sequence is not None:
```

Comment:
Why are we skipping over missing keys here (sequence, part) vs the Findaway-specific metadata, where missing keys are set to None? I'm not familiar with what a part vs a sequence is, but could this cause problems with parsing in the mobile client?

Reply:
I also don't know what these items mean, but it's always good to avoid a situation where the parser expects a key to be there and it's not there.

Reply:
👍